### ISSUE TITLE
False positive ProducerMessage.Results

### ISSUE THREAD

**Issue by yiminyuan** [akka/alpakka-kafka#1106]
**Versions**:
- Scala 2.12.11
- Akka 2.5.26
- Alpakka Kafka 2.0.2

I have the message loss problem if the Kafka producer meets `NotLeaderForPartitionException` when partitions got reassigned. Please check https://github.com/akka/alpakka-kafka/issues/438 for symptoms.

Let's say I have a pattern like this:
```scala
source
  .map { _ =>
    ProducerMessage.single(...)
  }
  .via(Producer.flexiFlow(producerSettings))
  .map {
    case ProducerMessage.Result(metadata, ProducerMessage.Message(...)) =>
      s"Successfully produced message for partition ${metadata.partition}, offset ${metadata.offset}: ${record.value}"
  }
  .runWith(...)
```
Then I may have some false positive logs for those messages that have been dropped. Let's say message A is lost and message B has been sent successfully, we could have logs like this:
```
Successfully produced message for partition 0, offset 1: A
Successfully produced message for partition 0, offset 1: B
```
The log for A is false positive. Notice that A and B have the **SAME** metadata.

I saw we have an issue at https://github.com/akka/alpakka-kafka/issues/1101 to handle the publication failure. I wonder if this could be a part of that improvement.

**Comment by ennru** [comment]
Thanks for the report. 
Interesting that the send callback is called as if it was successful.

**Comment by yiminyuan** [comment]
I figured out the root cause of this problem. If a partition reassignment causes leader reelection, then the record could be lost due to the default Kafka producer's setting: `acks=1`.

I drew a diagram for the scenario in this ticket (Assume the topic has 3 replicas A, B, C, and all of them are in-sync. So any of them could be elected as a new leader):
![01](https://user-images.githubusercontent.com/13373154/84212309-59ee3b00-aa72-11ea-8d32-50b00e6721c7.jpg)
![02](https://user-images.githubusercontent.com/13373154/84212316-5c509500-aa72-11ea-9f6e-dabe913692a0.jpg)
![03](https://user-images.githubusercontent.com/13373154/84212319-5eb2ef00-aa72-11ea-97a3-9d790c969930.jpg)
![04](https://user-images.githubusercontent.com/13373154/84212323-607cb280-aa72-11ea-9848-f38a5b08aa39.jpg)
![05](https://user-images.githubusercontent.com/13373154/84212325-62df0c80-aa72-11ea-87fe-4915ea0464bd.jpg)

I did a bunch of experiments by setting `acks=all` and I don't observe record is lost anymore due to leader reelection. However, it could produce duplicate messages. You may need to play around `enable.idempotence`  if the duplicate is intolerable.

Feel free to discuss more on this ticket or close this ticket if needed.

**Comment by ennru** [comment]
Thank you for digging into this and sharing your findings.

I agree it would be great to discuss the need for idempotence and acks. This does not only apply to Alpakka Kafka, though. But some pointers to these settings might be appropriate where the docs explain `ProducerSettings`.

**Comment by seglo** [comment]
I agree with @ennru that this issue is of a larger scope than just Alpakka Kafka (it's a Kafka producing corner case), but a PR to add it to the docs would be welcome.  I think a small mention in the `ProducerSettings` is warranted as @ennru already noted, and if you want to go to get into more detail then create a new page under the `Producer` section (not a sub heading, but a new page entirely, like how "[Subscription](https://doc.akka.io/docs/alpakka-kafka/current/subscription.html)" is a sub page of "Consumer") and add a more detailed account (including maybe the graphics?)

**Comment by yiminyuan** [comment]
Thank you @ennru and @seglo for your responses. I noticed that [Transactional Sink and Flow](https://doc.akka.io/docs/alpakka-kafka/current/transactions.html#transactional-sink-and-flow) did mention that `enable.idempotence=true` and `max.in.flight.requests.per.connection=1`, which implicitly implies `acks=all`. Hence, the record durability is guaranteed for the transactional case.

For the non-transactional case, the record durability is required for some cases. One of our use cases is that duplicate-record is fine but losing-record is intolerable. I think it would make sense to add a section to the [Producer](https://doc.akka.io/docs/alpakka-kafka/current/producer.html) page, and also add a pointer to this issue page to that section if other folks are interested in reading in more detail.

I am happy to open a PR but I have never contributed to the codebase. Should I sign the Lightbend CLA before opening a PR? Thank you!

**Comment by seglo** [comment]
@yiminyuan I think that would be a fine addition, but please put it in a new page under producer as I outlined in my earlier comment. I assume all the graphics are your own IP? 

You can view your updates to the docs with the `docs/previewSite` SBT task.

Yes, if you haven't already signed the CLA then you'll have to do that too.  If you've already done it in other akka projections then you should be good to go.

**Comment by yiminyuan** [comment]
IP stands for intellectual property? If so, yes! I drew those graphics by myself after some research. Thank you for the tips. I will open a PR and ask you to review.
